x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datatiers/NodesDataTiersUsageTransportAction.java

```diff
@@ -0,0 +1,226 @@
+    NodeDataTiersUsage> {
+
+    public static final ActionType<NodesResponse> TYPE = ActionType.localOnly("cluster:monitor/nodes/data_tier_usage");
+    private static final CommonStatsFlags STATS_FLAGS = new CommonStatsFlags().clear()
+        .set(CommonStatsFlags.Flag.Docs, true)
+        .set(CommonStatsFlags.Flag.Store, true);
+
+    private final IndicesService indicesService;
+
+    @Inject
+    public NodesDataTiersUsageTransportAction(
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        TransportService transportService,
+        IndicesService indicesService,
+        ActionFilters actionFilters
+    ) {
+        super(
+            TYPE.name(),
+            clusterService,
+            transportService,
+            actionFilters,
+            NodeRequest::new,
+            threadPool.executor(ThreadPool.Names.MANAGEMENT)
+        );
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected NodesResponse newResponse(NodesRequest request, List<NodeDataTiersUsage> responses, List<FailedNodeException> failures) {
+        return new NodesResponse(clusterService.getClusterName(), responses, failures);
+    }
+
+    @Override
+    protected NodeRequest newNodeRequest(NodesRequest request) {
+        return NodeRequest.INSTANCE;
+    }
+
+    @Override
+    protected NodeDataTiersUsage newNodeResponse(StreamInput in, DiscoveryNode node) throws IOException {
+        return new NodeDataTiersUsage(in);
+    }
+
+    @Override
+    protected NodeDataTiersUsage nodeOperation(NodeRequest nodeRequest, Task task) {
+        assert task instanceof CancellableTask;
+
+        DiscoveryNode localNode = clusterService.localNode();
+        NodeIndicesStats nodeIndicesStats = indicesService.stats(STATS_FLAGS, true);
+        ClusterState state = clusterService.state();
+        RoutingNode routingNode = state.getRoutingNodes().node(localNode.getId());
+        Map<String, NodeDataTiersUsage.UsageStats> usageStatsByTier = aggregateStats(routingNode, state.metadata(), nodeIndicesStats);
+        return new NodeDataTiersUsage(clusterService.localNode(), usageStatsByTier);
+    }
+
+    // For testing
+    static Map<String, NodeDataTiersUsage.UsageStats> aggregateStats(
+        RoutingNode routingNode,
+        Metadata metadata,
+        NodeIndicesStats nodeIndicesStats
+    ) {
+        if (routingNode == null) {
+            return Map.of();
+        }
+        Map<String, NodeDataTiersUsage.UsageStats> usageStatsByTier = new HashMap<>();
+        Set<String> localIndices = StreamSupport.stream(routingNode.spliterator(), false)
+            .map(routing -> routing.index().getName())
+            .collect(Collectors.toSet());
+        for (String indexName : localIndices) {
+            IndexMetadata indexMetadata = metadata.index(indexName);
+            String tier = indexMetadata.getTierPreference().isEmpty() ? null : indexMetadata.getTierPreference().get(0);
```

Comment:
I think we should guard against an empty `tier_preference`. 
`indexMetadata.getTierPreference().get(0)` could throw index out bounds exception

Reply:
I agree, I think this is covered by `indexMetadata.getTierPreference().isEmpty()`,  right? Is there an edge case I am missing?

Reply:
I think this is covered, but so I go ahead and merge this, but if I am indeed missing something I will fix it asap.

Reply:
Oh my eyes skipped that (Happy Friday 😬 )